src/Membership.sol

```diff
@@ -0,0 +1,427 @@
+import "openzeppelin-contracts/contracts/token/ERC20/IERC20.sol";
+import "openzeppelin-contracts/contracts/token/ERC20/utils/SafeERC20.sol";
+import "openzeppelin-contracts/contracts/utils/Context.sol";
+
+// The number of periods should be greater than zero
+error NumberOfPeriodsCantBeZero();
+
+// ETH Amount passed as value on the transaction is not correct
+error IncorrectAmount();
+
+// An eth value was assigned in the transaction and only tokens were expected
+error OnlyTokensAccepted();
+
+// The specified rate limit was not correct or within the expected limits
+error InvalidRateLimit();
+
+// It's not possible to acquire the rate limit due to exceeding the expected limits
+// even after attempting to erase expired memberships
+error ExceedMaxRateLimitPerEpoch();
+
+// This membership is not in grace period yet
+error NotInGracePeriod(uint256 idCommitment);
+
+// The sender is not the holder of the membership
+error NotHolder(uint256 idCommitment);
+
+// This membership cannot be erased (either it is not expired or not in grace period and/or not the owner)
+error CantEraseMembership(uint256 idCommitment);
+
+contract Membership {
+    using SafeERC20 for IERC20;
+
+    /// @notice Address of the Price Calculator used to calculate the price of a new membership
+    IPriceCalculator public priceCalculator;
+
+    /// @notice Maximum total rate limit of all memberships in the tree
+    uint32 public maxTotalRateLimitPerEpoch;
+
+    /// @notice Maximum rate limit of one membership
+    uint32 public maxRateLimitPerMembership;
+
+    /// @notice Minimum rate limit of one membership
+    uint32 public minRateLimitPerMembership;
+
+    /// @notice Membership billing period
+    uint32 public billingPeriod;
+
+    /// @notice Membership grace period
+    uint32 public gracePeriod;
+
+    /// @notice balances available to withdraw
+    mapping(address => mapping(address => uint256)) public balancesToWithdraw; // holder ->  token -> balance
+
+    /// @notice Total rate limit of all memberships in the tree
+    uint256 public totalRateLimitPerEpoch;
+
+    /// @notice List of registered memberships (IDCommitment to membership metadata)
+    mapping(uint256 => MembershipInfo) public members;
+
+    /// @notice The index of the next member to be registered
+    uint32 public commitmentIndex;
+
+    /// @notice track available indices that are available due to expired memberships being removed
+    uint32[] public availableExpiredIndices;
+
+    /// @dev Oldest membership
+    uint256 public head = 0;
+
+    /// @dev Newest membership
+    uint256 public tail = 0;
+
+    struct MembershipInfo {
+        /// @notice idCommitment of the previous membership
+        uint256 prev;
+        /// @notice idCommitment of the next membership
+        uint256 next;
+        /// @notice amount of the token used to acquire this membership
+        uint256 amount;
+        /// @notice numPeriods
+        uint32 numberOfPeriods;
```

Comment:
What's the period length in days and maximum number of periods we expect? depending on this we can change the variable type. If we assume billing periods are 30d, then an `uint8` would allow us to specify a max of 255 months = 21y, meaning that uint8 would be large enough for the purposes. It all depends on the values we choose for `billingPeriod` and maximum number of periods

Reply:
Note I'm storing the number of periods here, so when we extend the membership we calculate the new expiration date based on the same number of periods used originally (otherwise an user could register with a small number of period, and transfer the tokens associated to this period, and then extend for a much larger number of periods with no cost.

Reply:
Worth mentioning too that storing the number of periods incur in extra gas cost for the user

Reply:
Since the concept of billing periods was removed, these questions/comments are no longer relevant